datafusion/expr/src/planner.rs

```diff
@@ -0,0 +1,133 @@
+    /// We don't directly implement this in the logical plan's ['SqlToRel`]
+    /// because the sql code needs access to a table that contains execution-related types that can't be a direct dependency
+    /// of the sql crate (namely, the `CteWorktable`).
+    /// The [`ContextProvider`] provides a way to "hide" this dependency.
+    fn create_cte_work_table(
+        &self,
+        _name: &str,
+        _schema: SchemaRef,
+    ) -> Result<Arc<dyn TableSource>> {
+        not_impl_err!("Recursive CTE is not implemented")
+    }
+
+    /// Getter for a UDF description
+    fn get_function_meta(&self, name: &str) -> Option<Arc<ScalarUDF>>;
+    /// Getter for a UDAF description
+    fn get_aggregate_meta(&self, name: &str) -> Option<Arc<AggregateUDF>>;
+    /// Getter for a UDWF
+    fn get_window_meta(&self, name: &str) -> Option<Arc<WindowUDF>>;
+    /// Getter for system/user-defined variable type
+    fn get_variable_type(&self, variable_names: &[String]) -> Option<DataType>;
+
+    /// Get configuration options
+    fn options(&self) -> &ConfigOptions;
+
+    /// Get all user defined scalar function names
+    fn udf_names(&self) -> Vec<String>;
+
+    /// Get all user defined aggregate function names
+    fn udaf_names(&self) -> Vec<String>;
+
+    /// Get all user defined window function names
+    fn udwf_names(&self) -> Vec<String>;
+}
+
+/// This trait allows users to customize the behavior of the SQL planner
+pub trait UserDefinedPlanner {
+    /// Plan the binary operation between two expressions, returns OriginalBinaryExpr if not possible
+    fn plan_binary_op(
+        &self,
+        expr: BinaryExpr,
+        _schema: &DFSchema,
+    ) -> Result<PlannerSimplifyResult> {
+        Ok(PlannerSimplifyResult::OriginalBinaryExpr(expr))
+    }
+
+    /// Plan the field access expression, returns OriginalFieldAccessExpr if not possible
+    fn plan_field_access(
+        &self,
+        expr: FieldAccessExpr,
+        _schema: &DFSchema,
+    ) -> Result<PlannerSimplifyResult> {
+        Ok(PlannerSimplifyResult::OriginalFieldAccessExpr(expr))
+    }
+
+    // Plan the array literal, returns OriginalArray if not possible
+    fn plan_array_literal(
+        &self,
+        exprs: Vec<Expr>,
+        _schema: &DFSchema,
+    ) -> Result<PlannerSimplifyResult> {
+        Ok(PlannerSimplifyResult::OriginalArray(exprs))
+    }
+}
+
+pub struct BinaryExpr {
+    pub op: sqlparser::ast::BinaryOperator,
+    pub left: Expr,
+    pub right: Expr,
+}
+
+pub struct FieldAccessExpr {
+    pub field_access: GetFieldAccess,
+    pub expr: Expr,
+}
+
+pub enum PlannerSimplifyResult {
+    /// The function call was simplified to an entirely new Expr
+    Simplified(Expr),
+    /// the function call could not be simplified, and the arguments
+    /// are return unmodified.
+    OriginalBinaryExpr(BinaryExpr),
+    OriginalFieldAccessExpr(FieldAccessExpr),
+    OriginalArray(Vec<Expr>),
+}
```

Comment:
I think this might be easier to use if it were generic so that the callsites would know exactly the type of the returns Original

SO something like


```suggestion
pub enum PlannerSimplifyResult<T> {
    /// The structure was planned as an entirely new Expr by the planner
    Simplified(Expr),
    /// the planner did not handle planning the structure, and it is returned
    /// unmodified.
    OriginalBinaryExpr(T),
}
```